### PR TITLE
Add messaging layer

### PR DESCRIPTION
**Pull request by tcharding**
Adds a messaging layer similar to what we did in `thor`. Please note this PR duplicates the current happy path test into a new module then adds code to that module. 

With this applied the happy path function is even more horrendous than it was. Have not worked out how to make it easier to read yet. 

If this merges and the approach (as well as the cleanup) is acceptable to the team then we can remove the duplication. This is an attempt to not put others in merge hell while still having the freedom to do useful work as I see it. Open to suggestions.

### CODE REVIEW

File: xmr-btc/src/happy_path.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+pub async fn init_bitcoind(tc_client: &Cli) -> Bitcoind<'_> {
+    let bitcoind = Bitcoind::new(tc_client, "0.19.1").expect("failed to create bitcoind");
+    let _ = bitcoind.init(5).await;
+
+    bitcoind
+}
+
+/// Create two mock `Transport`s which mimic a peer to peer connection between
+/// two parties, allowing them to send and receive `thor::Message`s.
+pub fn make_transports() -> (Transport, Transport) {
```
Comment (bonomat) [inline xmr-btc/src/happy_path.rs:30]: This looks similar to what @rishflab did in his PR in https://github.com/comit-network/xmr-btc-swap/pull/2 
see https://github.com/comit-network/xmr-btc-swap/blob/a5e9d12a6c6c489eccd6de6dd4ae85da38f2e60f/xmr-btc/tests/e2e.rs#L41-L60

File: xmr-btc/src/happy_path.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    async fn receive_message(&mut self) -> Result<Message> {
+        let msg = self
+            .receiver
+            .next()
+            .await
+            .ok_or_else(|| anyhow!("failed to receive message"))?;
+
+        Ok(msg)
+    }
+}
```
Comment (bonomat) [inline xmr-btc/src/happy_path.rs:74]: This seems also to be very similar to what @rishflab did in his PR #2 in https://github.com/comit-network/xmr-btc-swap/blob/b6df715b1bdef81f20333145e1705cbdcf0e4fb8/xmr-btc/tests/harness/transport.rs#L16-L55

The difference is that @rishflab made a difference in the message types and uses two different enums, one for Alice's messages and one for Bob's messages. This should prevent anyone from sending the wrong message.

File: xmr-btc/src/happy_path.rs (hunk 1 above)
Comment (tcharding) [reply xmr-btc/src/happy_path.rs:30]: Yes except he puts two methods on a single trait - a single method per trait is cleaner. Also his traits are never used, writing these traits takes about 2 minutes of cut and pasting from `thor`. I do not see this as an issue.

Comment (bonomat) [reply xmr-btc/src/happy_path.rs:30]: True, he uses a single trait. That can be refactored easily. 

It looks like the traits are used in the `next_state` function for Alice and Bob 

https://github.com/comit-network/xmr-btc-swap/blob/b6df715b1bdef81f20333145e1705cbdcf0e4fb8/xmr-btc/src/alice.rs#L34-L35

and 

https://github.com/comit-network/xmr-btc-swap/blob/b6df715b1bdef81f20333145e1705cbdcf0e4fb8/xmr-btc/src/bob.rs#L37-L41

Comment (tcharding) [reply xmr-btc/src/happy_path.rs:30]: > Tthe traits are used in the next_state function for Alice and Bob

The call sites you reference are using the function parameter that is bound by the trait. This code is not using the implementation of the trait that we are discussing.

**Pull request closed without merging**